DOCSP-48683 Refine Connection Pools page (#1067)
mongodb/docs-node: source/connect/connection-options/connection-pools.txt

Reviewer: Trying to think of a new title for page. Perhaps "Manage Connections in a Connection Pool"?

Author: Was also thinking about this! I don't think this page made it into the spreadsheet, it's not in all the drivers docs. I think that title makes sense.

Author: Or maybe "Manage Connections with Connection Pools"

Reviewer: I like "Manage Connections with Connection Pools", great suggestion!

.. _node-faq-connection-pool:

> Configuring Connection Pools

Reviewer: ```suggestion
Configure Connection Pools
```

--------------------------

Every ``MongoClient`` instance has a built-in connection pool for each server
in your MongoDB topology. Connection pools open sockets on demand to

> support concurrent requests to MongoDB in your application.

Reviewer 2: This is only true if there isn't a minPoolSize.  I'm not sure the best way to phrase this - something like this seems clunky:

"Connection pools open sockets on demand to support concurrent requests to MongoDB in your application **when minPoolSize is not configured**."

Author: Gotcha, this nuance is good to know will work on re-wording this.

   * - Setting
     - Description

   * - ``maxPoolSize``
     - | The maximum number of concurrent connections that the pool maintains.
         If the number of in-use connections to a server reaches the specified
         value, the next request to that server waits until a connection becomes
         available.
       |
       | **Default**: ``100``

   * - ``maxConnecting``
     - | The maximum number of connections that each pool can establish
         concurrently.

   * - ``minPoolSize``
     - | The minimum number of concurrent connections that the pool maintains.
       |
       | **Default**: ``0``

   * - ``maxIdleTimeMS``
     - | The maximum number of milliseconds that a connection can remain idle in
         the pool.
       |
       | **Default**: ``0`` (no limit)

   * - ``waitQueueTimeoutMS``
     - | The maximum number of milliseconds that a request can wait for a socket
         to become available.
       |
       | **Default**: ``0`` (no limit)

.. _node-connection-pool-max-pool-size:

maxPoolSize
~~~~~~~~~~~

In addition to the sockets needed to support your application's requests,
each ``MongoClient`` instance opens two more sockets per server

> in your MongoDB topology for monitoring the server's state.

Reviewer 2: Not sure how precise we want to be with the language here, but the number of connections opened by driver depends on how the driver is monitoring.  In most cases, it is 2 per server, but that isn't true if:

1. the user is using a load balancer (no monitoring is done if there is a load balancer configured)
2. the user is using a pre-4.4 server (then we only use 1 connection per server)
3. the user is operating in a FAAS environment, like AWS Lambda (then there will only be one connection per server)

Maybe we could say "up to two connection per server"?

For example, a client connected to a three-node replica set opens six
monitoring sockets. If the application uses the default setting for

> ``maxPoolSize`` and only queries the primary (default) node, then

Reviewer 2: Monitoring sockets don't actually live in the connection pools.  Maybe we can use phrasing like "the driver might open 106 sockets and there will be 100 connections in the connection pool"?  Or something similar.

instance:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Setting
     - Description

   * - ``maxPoolSize``
     - | The maximum number of concurrent connections that the pool maintains.
         If the number of in-use connections to a server reaches the specified
         value, the next request to that server waits until a connection becomes
         available.
       |
       | **Default**: ``100``

   * - ``maxConnecting``
     - | The maximum number of connections that each pool can establish
         concurrently.

   * - ``minPoolSize``
     - | The minimum number of concurrent connections that the pool maintains.
       |
       | **Default**: ``0``

   * - ``maxIdleTimeMS``
     - | The maximum number of milliseconds that a connection can remain idle in
         the pool.
       |
       | **Default**: ``0`` (no limit)

   * - ``waitQueueTimeoutMS``
     - | The maximum number of milliseconds that a request can wait for a socket
         to become available.
       |
       | **Default**: ``0`` (no limit)

.. _node-connection-pool-max-pool-size:

maxPoolSize
~~~~~~~~~~~

In addition to the sockets needed to support your application's requests, each

> ``MongoClient`` instance opens up to two or more connections per server in your

Reviewer 2: ```suggestion
``MongoClient`` instance opens up to two connections per server in your
```

It is always either 0, 1 or 2 connections per server

MongoDB topology for monitoring the server's state.

For example, a client connected to a three-node replica set opens six monitoring
sockets. If the application uses the default setting for ``maxPoolSize`` and
only queries the primary (default) node, then there can be at most ``106`` open
sockets and ``100`` connections in the connection pool. If the application uses
a :ref:`read preference <read-preference>` to query the secondary nodes, those
connection pools grow and there can be ``306`` total connections including the
open monitoring sockets.

To support high numbers of concurrent MongoDB requests
within one process, you can increase ``maxPoolSize``.

The following code creates a ``MongoClient`` instance with a maximum connection
pool size of ``200`` by specifying the ``maxPoolSize`` option in the
``options`` object:

.. code-block:: javascript

   const { MongoClient } = require('mongodb');

   const uri = '<connection-string>';
   const client = new MongoClient(uri, {
     maxPoolSize: 200
   });

.. _node-connection-pool-max-connecting:

maxConnecting
~~~~~~~~~~~~~

> Connection pools rate-limit of connection establishment. The ``maxConnecting``

Reviewer 2: ```suggestion
Connection pools rate-limit connection establishment. The ``maxConnecting``
```

To support high numbers of concurrent MongoDB requests
within one process, you can increase ``maxPoolSize``.

The following code creates a ``MongoClient`` instance with a maximum connection
pool size of ``50`` by specifying the ``maxPoolSize`` option in the
``options`` object:

.. code-block:: javascript

   const { MongoClient } = require('mongodb');

   const uri = '<connection-string>';
   const client = new MongoClient(uri, {

>      maxPoolSize: 50

Reviewer 2: suggest using a number higher than the default, given that the preceding sentence mentions increasing maxPoolSize.

   });

.. _node-connection-pool-max-connecting:

maxConnecting
~~~~~~~~~~~~~

> Connection pools are rate-limited. The ``maxConnecting`` option

Reviewer 2: ```suggestion
Connection pools rate-limit connection establishment. The ``maxConnecting`` option
```

> - The driver's ability to reuse existing connections improves due to
>   rate-limits on connection creation.

Reviewer 2: ```suggestion
```

I'm not sure what this bullet point is actually getting at, I'd suggest just removing it.

The following code creates a ``MongoClient`` instance with a maximum number of

> ``2`` connections by specifying the ``maxConnecting`` option in the ``options``

Reviewer 2: ```suggestion
``2`` connections to be established concurrently per pool by specifying the ``maxConnecting`` option in the ``options``
```

object:

.. code-block:: javascript

   const { MongoClient } = require('mongodb');

   const uri = '<connection-string>';
   const client = new MongoClient(uri, {
      maxConnecting: 2
   });

.. _node-connection-pool-min-pool-size:

minPoolSize
~~~~~~~~~~~

> You can set the minimum number of concurrent connections to each server with the

Reviewer 2: ```suggestion
You can set the minimum number of connections to each server with the
```

> ``minPoolSize`` option. The driver initializes the connection pool with the
> number of sockets set by the option. If sockets are closed, causing the total

Reviewer 2: ```suggestion
``minPoolSize`` option. The driver will ensure there are always at least minPoolSize connections in the connection pool. If sockets are closed, causing the total
```

just a nit: the driver always initializes with 0 connections

> ``minPoolSize`` option. The driver initializes the connection pool with this
> number of sockets. If sockets are closed, causing the total number of sockets

Reviewer: ```suggestion
``minPoolSize`` option. The driver initializes the connection pool with the
number of sockets set by the option. If sockets are closed, causing the total number of sockets
```
S: For clarity

directly terminate any ongoing operations. 

.. note::

   The ``MongoClient.close()`` method does close existing sessions and
   transactions, which might indirectly affect the behavior of ongoing
   operations and open cursors.

Avoid Socket Timeouts
---------------------

Having a large connection pool does not always reduce reconnection
requests. Consider the following example scenario:

- An application has a connection pool size of 5 sockets and has the
  ``socketTimeoutMS`` option set to 5000 milliseconds. 
- Operations occur, on average, every 3000 milliseconds, and reconnection
  requests are frequent. 
- Each socket times out after 5000 milliseconds, which means that all sockets
  must do something during those 5000 milliseconds to avoid closing.

In this scenario, each socket times out after 5000 milliseconds, requiring
activity within this timeout period to avoid closure. However, one message every
3000 milliseconds isn't enough to keep all sockets active, causing several of
them to time out.

To avoid excessive socket timeouts, reduce the number of connections that the
driver can maintain in the connection pool by specifying the ``maxPoolSize``

> option. To learn how to set this option, see the

Reviewer: ```suggestion
option. To learn how to set the ``maxPoolSize`` option, see the
```
S: For clarity

:ref:`node-connection-pool-max-pool-size` section.

API Documentation
-----------------

For more information about creating a ``MongoClient`` object with the 
{+driver-short+} and specifying options, see the following API documentation: 

- `MongoClient <{+api+}/classes/MongoClient.html>`__

> - `MongoClientOptions <{+api+}/interfaces/MongoClientOptions.html>`__

Reviewer: S: Since the Options page is so long, it would be nice to include the specific links to each of the relevant options.
